### PR TITLE
Add location-based history tracking on Database3

### PR DESCRIPTION
**Pull request by youngmit**
This adds the `getHistoryByLocation()` and `getHistoriesByLocation()`
methods to the Database3 class. These are quite similar to their
non-location based siblings, but rather than following an ArmiObject in
the database by serialNum, it does a little extra work to figure out
each object's location, then pull data from the object that is actually
at that location for each time node requested. Other changes:
 - Added some type annotation stuff
 - Added `computeParents()` classmethod to Layout to help make sense of
 the hierarchy without having to fully reconstruct.
 - Added a `getAncestorWithDistance()` method to ArmiObject. This is
 useful for determining not only the ancestor object, but how many
 layers above it happens to be.
 - Moved some methods that should obviously be Composite methods out of
 ArmiObject. There are plenty more that are miscategorized and still
 need moved, but that should be its own thing.

### CODE REVIEW

**Comment by youngmit** [comment]
Working on cleaning up some docs, and removing the `anchor` kwarg. It doesn't actually work right now, and would take some more head scratching to figure out properly.

File: armi/bookkeeping/db/database3.py [hunk 1]
```diff
@@ -1776,6 +2035,65 @@ def writeToDB(self, h5group):
             runLog.error("Failed to create datasets in: {}".format(h5group))
             raise
 
+    @staticmethod
+    def computeAncestors(serialNum, numChildren, depth=1) -> List[Optional[int]]:
+        """
+        Return a list containing the serial number of the parent corresponding to each
+        object at the given depth.
```
Comment (ntouran) [inline armi/bookkeeping/db/database3.py:2042]: can you explain that slightly more? I see a serialNum input that is hard to reconcile with this description quickly

File: armi/reactor/composites.py [hunk 2]
```diff
@@ -1392,6 +1392,26 @@ def getAncestor(self, fn):
         else:
             return self.parent.getAncestor(fn)
 
+    def getAncestorWithDistance(
```
Comment (ntouran) [inline armi/reactor/composites.py:1395]: Maybe `getAncestorAndDistance`? With distance seems to imply to me that you're specifying a distance .

File: armi/bookkeeping/db/database3.py [hunk 3]
```diff
@@ -1161,11 +1196,199 @@ def _readParams(h5group, compTypeName, comps):
                         " (possibly loading from old DB)\n"
                     )
 
+    def getHistoryByLocation(
+        self,
+        comp: ArmiObject,
```
[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1201]: Astute! I originally was thinking about doing it that way, but changed my mind for a couple of reasons:
 - To pass just a location would require explicitly passing the anchor object, or doing some pretty weird grid magic that is pretty difficult to get right.
 - Having the object itself helps make sense of the location's relationship to another actual object, which is the most efficient way to figure out which indices from each time step that we are interested in.
 - Typically what a caller probably has is an object anyways, so digging for its spatial locator just to go back into the reactor tree to start traversing seemed a bit odd.

I will make it more clear in the docs. If we find that there is a good reason to want to do it with just a spatial locator, we can do the extra grid stuff to make it work, and change the type of the argument to be something like `Union[ArmiObject, IndexLocation]` or similar

File: armi/bookkeeping/db/database3.py [hunk 4]
```diff
@@ -1161,11 +1196,199 @@ def _readParams(h5group, compTypeName, comps):
...
+        comps: Sequence[ArmiObject],
+        params: Optional[List[str]] = None,
+        timeSteps: Optional[Sequence[Tuple[int, int]]] = None,
+    ) -> Histories:
+        """
+        Get the parameter histories at specific locations.
+
+        This has a number of limitations, which should in practice not be too limiting:
+         - The passed locations must be IndexLocations. This type of operation doesn't
+           make much sense otherwise.
```
[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1224]: yup. wrote that when i was still taking locations as argument

File: armi/bookkeeping/db/database3.py [hunk 5]
```diff
@@ -1161,11 +1196,199 @@ def _readParams(h5group, compTypeName, comps):
...
+            histories at. ArmiObjects are passed, rather than locations, because this
+            makes it easier to figure out things related to layout.
+        params : List of str, optional
+            The parameter names for the parameters that we want the history of. If None,
+            all parameter history is given
+        timeSteps : List of (cycle, node) tuples, optional
+            The time nodes that you want history for. If None, all available time nodes
+            will be returned.
+        """
+        if self.versionMinor < 4:
```
[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1246]: I think by definition a new major version would necessitate a new module, with no shared implementation. By the time this function is being called at all, we already know we are dealing with a major version 3 database

[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1246]: I neglected to make this intent clear in the module docstring, so I'll expound upon that there

[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1246]: Looks like there is some discussion, but a level above in `db.__init__`. I'll add a link

File: armi/bookkeeping/db/database3.py [hunk 6]
```diff
@@ -1161,11 +1196,199 @@ def _readParams(h5group, compTypeName, comps):
...
+            lLocation = layout.location
+            # filter for objects that live under the desired ancestor and at a desired
+            # location
+            # TODO: There might be a numpy way of doing this faster, were we to treat
+            # the locations as a numpy array. The elements are tuple of int, tuple of
+            # float, or sometimes even None, as determined by the pack/unpackLocations
+            # implementations, so it might not be possible, let alone trivial to do
+            # this. One approach could be to go back to the locations in their raw
+            # HDF5 form, then list index into that, along with locationType, and
+            # re-unpack them. 🤔
```
[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1313]: Yeah, decided to have fun with it

File: armi/bookkeeping/db/database3.py [hunk 7]
```diff
@@ -1343,74 +1566,115 @@ def _packLocation(
...
+        else:
+            raise ValueError(f"Invalid location type: {loc}")
+
+        locTypes.append(locationType)
+        locData.extend(locDatum)
+
+    return locTypes, locData
+
+
+def _packLocationsV2(
```
[the remark this answers is not in the document]
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:1609]: Added blurb referring to module level revision notes

File: armi/bookkeeping/db/database3.py (hunk 1 above)
Comment (youngmit) [reply armi/bookkeeping/db/database3.py:2042]: I at least added argument doc strings and more info on `depth`.

File: armi/reactor/composites.py (hunk 2 above)
Comment (youngmit) [reply armi/reactor/composites.py:1395]: Good call

**Comment by ntouran** [comment]
looks good except for the failing unit test

**Comment by youngmit** [comment]
Bah! Missed a find/replace. If tests pass, ill merge

**Pull request merged**